Adds backend for GCS via service account and the official GCP node client.

nickbauman opened the pull request.

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,152 @@
+// Google Cloud Storage backend for uploadfs. See also
+// local.js.
+
+var {Storage} = require('@google-cloud/storage');
```
Please stick to idioms supported back to Node 6. `const` and `let` are fine but object destructuring is not supported.

boutell replied:
i.e. `var Storage = require('@google-cloud/storage').Storage;`

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,152 @@
+  var self = {
+    init: function (options, callback) {
+      if (!(process.env.GOOGLE_APPLICATION_CREDENTIALS)) {
+        throw "GOOGLE_APPLICATION_CREDENTIALS not set in env, cannot proceed";
```
Should pass this to `callback` as first argument, rather than using `throw`.

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,152 @@
+            destination: cleanPath,
+            gzip: true,
+            metadata: {
+              cacheControl: cacheControl,
```
Do the capitalization differences between cacheControl and ContentType make sense? Honestly not sure.

nickbauman replied:
I took this directly from the Google documentation for the client. I will find a definitive answer and report back.

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,152 @@
+            callback(reason);
+          });
+        },
+        function(apiResponse, __) {
```
This works, but it should have its own callback, and the final `callback` should be passed as the final argument to `async.waterfall`, for consistency with the expectations of folks familiar with the module.

boutell replied:
I don't mind cleaning this up.

nickbauman replied:
Oops, you're right. I can fix this as I do the other changes you requested :)

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,152 @@
+              }, function (err) {
+                callback(err);
+              });
+            }).catch(function (reason) {
```
This works, the promise chaining could be a little cleaner, this can be a flat chain of calls with just one catch() at the end and no use of a second argument to then() for readability. I can take care of that if you like, really appreciate this contribution.

nickbauman replied:
I can do that in the course of doing the other changes you requested, Thomas. :)

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,152 @@
+    },
+
+    copyOut: function(path, localPath, options, callback) {
+      var mergedOptions = { destination: cleanKey(localPath), ...options };
```
Unfortunately the "rest syntax" is also not a Node 6 compatible thing, I'll fix.

nickbauman replied:
I can remove the use of the spread operator, it's NBD.

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,152 @@
+    },
+
+    disable: function(path, callback) {
+      client.bucket(bucketName).file(cleanKey(path)).makePrivate({}, callback);
```
Glad Google makes this part easy like S3 does.

boutell commented on test/gcs.js:
```diff
@@ -0,0 +1,169 @@
+  gcsOptions.tempPath = tempPath;
+  gcsOptions.params = {
+    Bucket: gcsOptions.bucket,
+    ACL: 'public-read'
```
Is this a requirement that needs to be in the documentation? What does it mean to set the ACL at this level rather than via makePublic one file at a time - is that needed?

boutell replied:
Does this mean the public can enumerate everything in the bucket?

nickbauman replied:
TL;DR, yes to both questions. 

I took the s3 tests "at their word" by copying them and making my version of the backend obey the same assertions. One of them was to assert that putting a resource into the remote bucket made it readable by an ordinary http client with no special access control. The GCS bucket I used for my testing was not public, only individual resources were made that way, so they should not be "enumeratable", but if you know the resource is there, it should be readable by anyone.

Let me know if you think this is reasonable.

nickbauman replied:
On second thought I think this part was slavishly copied from the s3 test. It has no effect on the ACLs on GCS, which is why I had to make it public when you `copyIn` elsewhere. I'll remove this. It's a canard.

nickbauman commented:
> One question though: the GCS options passed to the tests set an ACL for the bucket. Is that necessary? Does it need to be in the how-to documentation too?

I think it's a canard left over from the s3 tests and does nothing in GCS. I will remove.

boutell commented:
Sounds great - I'll wait for these updates. Really appreciate the
contribution.

Please do reach out to tom@punkave.com, I'd love to hear more about what
you're doing with Apostrophe and/or Uploadfs.

On Mon, Feb 18, 2019 at 7:57 PM Nick Bauman <notifications@github.com>
wrote:

> One question though: the GCS options passed to the tests set an ACL for
> the bucket. Is that necessary? Does it need to be in the how-to
> documentation too?
>
> I think it's a canard left over from the s3 tests and does nothing in GCS.
> I will remove.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/punkave/uploadfs/pull/48#issuecomment-464936464>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAB9fT0Mlwn2zshDhBu1g4XRwlsktWJjks5vO0vcgaJpZM4a-huy>
> .
>


-- 


*Thomas Boutell, Chief Software Architect*
P'unk Avenue | (215) 755-1330  |  punkave.com

nickbauman commented:
I think I've addressed your issues. I also was able to eliminate the need for the `async.waterfall` entirely in making resources publicly readable by a more careful reading of the API documentation.

nickbauman commented:
Oh, almost forgot! I did a rebase and force push. Hope that's OK? 😬

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,132 @@
+  const self = {
+    init: function (options, callback) {
+      if (!(process.env.GOOGLE_APPLICATION_CREDENTIALS)) {
+        callback("GOOGLE_APPLICATION_CREDENTIALS not set in env, cannot proceed");
```
Needs return statement or it'll fall through and do both.

boutell commented on lib/storage/gcs.js:
```diff
@@ -0,0 +1,132 @@
+      } else {
+        contentTypes = defaultTypes;
+      }
+      defaultTypes = require(__dirname + "/contentTypes.js");
```
Duplication of previous few lines.
